diff -burN vmhgfs-only.a/inode.c vmhgfs-only.b/inode.c
--- vmhgfs-only.a/inode.c	2015-01-07 15:51:14.000000000 -0800
+++ vmhgfs-only.b/inode.c	2015-03-09 09:14:34.949945519 -0700
@@ -1922,7 +1922,12 @@
                            p,
 #endif
                            &inode->i_dentry,
-                           d_alias) {
+#if defined(VMW_D_ALIAS_3181) || LINUX_VERSION_CODE >= KERNEL_VERSION(3, 18, 1)
+                           d_u.d_alias
+#else
+                           d_alias
+#endif
+                           ) {
          int dcount = hgfs_d_count(dentry);
          if (dcount) {
             LOG(4, ("Found %s %d \n", dentry->d_name.name, dcount));
@@ -1975,7 +1980,13 @@
       /* Find a dentry with valid d_count. Refer bug 587879. */
       list_for_each(pos, &inode->i_dentry) {
          int dcount;
-         struct dentry *dentry = list_entry(pos, struct dentry, d_alias);
+         struct dentry *dentry = list_entry(pos, struct dentry,
+#if defined(VMW_D_ALIAS_3181) || LINUX_VERSION_CODE >= KERNEL_VERSION(3, 18, 1)
+         d_u.d_alias
+#else
+         d_alias
+#endif
+         );
          dcount = hgfs_d_count(dentry);
          if (dcount) {
             LOG(4, ("Found %s %d \n", (dentry)->d_name.name, dcount));
diff -burN vmhgfs-only.a/Makefile.kernel vmhgfs-only.b/Makefile.kernel
--- vmhgfs-only.a/Makefile.kernel	2015-01-07 15:51:14.000000000 -0800
+++ vmhgfs-only.b/Makefile.kernel	2015-03-09 09:00:54.221027200 -0700
@@ -35,6 +35,7 @@
 EXTRA_CFLAGS += $(call vm_check_build, $(AUTOCONF_DIR)/statfs1.c,, -DVMW_STATFS_2618)
 EXTRA_CFLAGS += $(call vm_check_build, $(AUTOCONF_DIR)/inode1.c,, -DVMW_INODE_2618)
 EXTRA_CFLAGS += $(call vm_check_build, $(AUTOCONF_DIR)/dcount.c,, -DVMW_DCOUNT_311)
+EXTRA_CFLAGS += $(call vm_check_build, $(AUTOCONF_DIR)/d_alias.c,, -DVMW_D_ALIAS_3181)

 MODPOST_VMCI_SYMVERS := $(wildcard $(MODULEBUILDDIR)/VMwareVMCIModule.symvers)

diff -burN vmhgfs-only.a/shared/autoconf/d_alias.c vmhgfs-only.b/shared/autoconf/d_alias.c
--- vmhgfs-only.a/shared/autoconf/d_alias.c	1969-12-31 16:00:00.000000000 -0800
+++ vmhgfs-only.b/shared/autoconf/d_alias.c	2015-03-09 09:00:54.221027200 -0700
@@ -0,0 +1,21 @@
+/*********************************************************
+ * Copyright (C) 2015 Ross Smith II. MIT Licensed.
+ *
+ *********************************************************/
+
+#include "compat_version.h"
+#include "compat_autoconf.h"
+
+#include <linux/dcache.h>
+
+/*
+ * After 3.18.1, the dentry d_alias field was moved to struct d_u.
+ *
+ * This test will fail on a kernel with such a patch.
+ */
+void test(void)
+{
+   struct dentry dentry;
+
+   dentry.d_alias = NULL;
+}
