--- vmhgfs-only/inode.c.orig	2014-11-17 18:07:13.000000000 -0800
+++ vmhgfs-only/inode.c	2015-01-12 09:25:19.801902953 -0800
@@ -1922,7 +1922,11 @@
                            p,
 #endif
                            &inode->i_dentry,
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3, 18, 1) && !defined(__GENKSYMS__) && !defined(D_ALIAS_IS_A_MEMBER_OF_UNION_D_U)
                            d_alias) {
+#else
+                           d_u.d_alias) {
+#endif
          int dcount = hgfs_d_count(dentry);
          if (dcount) {
             LOG(4, ("Found %s %d \n", dentry->d_name.name, dcount));
@@ -1975,7 +1979,11 @@
       /* Find a dentry with valid d_count. Refer bug 587879. */
       list_for_each(pos, &inode->i_dentry) {
          int dcount;
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3, 18, 1) && !defined(__GENKSYMS__) && !defined(D_ALIAS_IS_A_MEMBER_OF_UNION_D_U)
          struct dentry *dentry = list_entry(pos, struct dentry, d_alias);
+#else
+         struct dentry *dentry = list_entry(pos, struct dentry, d_u.d_alias);
+#endif
          dcount = hgfs_d_count(dentry);
          if (dcount) {
             LOG(4, ("Found %s %d \n", (dentry)->d_name.name, dcount));
