diff -ru vmhgfs-only.orig/inode.c vmhgfs-only/inode.c
--- vmhgfs-only.orig/inode.c	2014-11-18 03:07:13.000000000 +0100
+++ vmhgfs-only/inode.c	2014-12-23 22:47:15.783855237 +0100
@@ -1922,7 +1922,7 @@
                            p,
 #endif
                            &inode->i_dentry,
-                           d_alias) {
+                           d_u.d_alias) {
          int dcount = hgfs_d_count(dentry);
          if (dcount) {
             LOG(4, ("Found %s %d \n", dentry->d_name.name, dcount));
