--- vmhgfs-only/inode.c.orig	2014-12-31 16:04:41.000000000 +0000
+++ vmhgfs-only/inode.c	2014-12-31 16:49:00.597010246 +0000
@@ -1893,7 +1893,7 @@
                            p,
 #endif
                            &inode->i_dentry,
-                           d_alias) {
+                           d_u.d_alias) {
          int dcount = compat_d_count(dentry);
          if (dcount) {
             LOG(4, ("Found %s %d \n", dentry->d_name.name, dcount));
